File: frontend/src/api/designPromotions.js
import { request, unwrapDetail } from "./client.js";
import {
  buildCandidateSelectionPayload,
  buildComparisonCreatePayload,
  buildPublishPayload,
  buildReleaseDecisionPayload,
  buildReleaseRequestPayload,
  buildRollbackPayload,
  validateDesignComparisonEnvelope,
  validateDesignPublishResult,
  validateDesignReleaseDecision,
  validateDesignReleaseRequest,
  validateDesignRollbackResult,
  validateDesignSelection,
  validateDesignPathId,
} from "../utils/designPromotionCore.js";

function pathId(value, field) {
  return encodeURIComponent(validateDesignPathId(value, field));
}

export function sensitiveCandidateRoleAxisMessage(err) {
  if (err?.status !== 403) return null;
  const detail = unwrapDetail(err.detail);
  if (!detail || typeof detail !== "object" || Array.isArray(detail)) return null;
  const keys = Object.keys(detail).sort();
  if (keys.length !== 2 || keys[0] !== "code" || keys[1] !== "message") return null;
  if (
    detail.code !== "sensitive_candidate_requires_role_axis" ||
    typeof detail.message !== "string" ||
    detail.message.trim() === ""
  ) {
    return null;
  }
  return detail.message;
}

export async function createDesignComparison(input) {
  const response = await request("/api/design-comparisons", {
    method: "POST",
    json: buildComparisonCreatePayload(input),
  });
  return validateDesignComparisonEnvelope(response);
}

export async function getDesignComparison(comparisonId) {
  const response = await request(
    `/api/design-comparisons/${pathId(comparisonId, "comparison_id")}`,
    { cache: "no-store" },
  );
  return validateDesignComparisonEnvelope(response);
}

export async function submitDesignSelection(comparisonId, input) {
  const response = await request(
    `/api/design-comparisons/${pathId(comparisonId, "comparison_id")}/selection`,
    { method: "POST", json: buildCandidateSelectionPayload(input) },
  );
  return validateDesignSelection(response);
}

export async function createDesignReleaseRequest(input) {
  const response = await request("/api/design-release-requests", {
    method: "POST",
    json: buildReleaseRequestPayload(input),
  });
  return validateDesignReleaseRequest(response);
}

export async function decideDesignReleaseRequest(releaseRequestId, input) {
  const response = await request(
    `/api/design-release-requests/${pathId(releaseRequestId, "release_request_id")}/decision`,
    { method: "POST", json: buildReleaseDecisionPayload(input) },
  );
  return validateDesignReleaseDecision(response);
}

export async function publishDesignRelease(releaseRequestId, input) {
  const response = await request(
    `/api/design-release-requests/${pathId(releaseRequestId, "release_request_id")}/publish`,
    { method: "POST", json: buildPublishPayload(input) },
  );
  return validateDesignPublishResult(response);
}

export async function rollbackDesignRelease(releaseRequestId, input) {
  const response = await request(
    `/api/design-release-requests/${pathId(releaseRequestId, "release_request_id")}/rollback`,
    { method: "POST", json: buildRollbackPayload(input) },
  );
  return validateDesignRollbackResult(response);
}

File: backend/app/api/design_promotions.py
"""Authenticated P2.8 comparison, release, publication, and rollback API.

Every actor is derived from the authenticated request principal.  Client
payloads contain no signer field, and publication is available only through an
explicit synchronous request carrying strict confirmation and exact hashes.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal

from fastapi import APIRouter, HTTPException, Path as ApiPath, Request, Response

from ..design_promotion.contracts import (
    Actor,
    ComparisonCreate,
    PublishRequest,
    ReleaseDecisionRequest,
    ReleaseRequestCreate,
    RollbackRequest,
    SelectionRequest,
)
from ..design_promotion.service import (
    DesignPromotionError,
    DesignPromotionService,
    PngFrame,
    SensitiveCandidateRequiresRoleAxis,
)


router = APIRouter(prefix="/api", tags=["design-promotions"])

ComparisonId = Annotated[
    str, ApiPath(pattern=r"^comparison_[a-f0-9]{32}$", max_length=43)
]
ReleaseRequestId = Annotated[
    str, ApiPath(pattern=r"^release_[a-f0-9]{32}$", max_length=40)
]
FrameId = Annotated[str, ApiPath(pattern=r"^frame_[a-f0-9]{32}$", max_length=38)]


def _service(request: Request) -> DesignPromotionService:
    service = getattr(request.app.state, "design_promotion_service", None)
    if not isinstance(service, DesignPromotionService):
        raise RuntimeError("design promotion service is not assembled")
    return service


def _actor(request: Request) -> Actor:
    principal = request.state.user
    return Actor(
        username=principal["username"], display_name=principal["display_name"]
    )


def _raise_http(exc: DesignPromotionError) -> None:
    if isinstance(exc, SensitiveCandidateRequiresRoleAxis):
        raise HTTPException(
            status_code=exc.status_code,
            detail={"code": exc.code, "message": str(exc)},
        ) from exc
    raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.post("/design-comparisons", status_code=201)
def create_design_comparison(
    body: ComparisonCreate, request: Request, response: Response
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).create_comparison(body, actor=_actor(request))
    except DesignPromotionError as exc:
        _raise_http(exc)


@router.get("/design-comparisons/{comparison_id}")
def get_design_comparison(
    comparison_id: ComparisonId, request: Request, response: Response
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).get_comparison(comparison_id)
    except DesignPromotionError as exc:
        _raise_http(exc)


@router.get(
    "/design-comparisons/{comparison_id}/frames/{frame_id}/{side}.png",
    response_class=Response,
)
def get_design_comparison_frame(
    comparison_id: ComparisonId,
    frame_id: FrameId,
    side: Literal["current", "candidate"],
    request: Request,
) -> Response:
    try:
        frame: PngFrame = _service(request).get_comparison_frame(
            comparison_id, frame_id, side
        )
    except DesignPromotionError as exc:
        _raise_http(exc)
    return Response(
        content=frame.content,
        media_type=frame.media_type,
        headers={
            "Cache-Control": "no-store",
            "Content-Disposition": "inline",
            "Content-Security-Policy": "default-src 'none'; sandbox",
            "X-Content-Type-Options": "nosniff",
        },
    )


@router.post("/design-comparisons/{comparison_id}/selection")
def decide_design_candidate(
    comparison_id: ComparisonId,
    body: SelectionRequest,
    request: Request,
    response: Response,
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).decide_candidate(
            comparison_id, body, actor=_actor(request)
        )
    except DesignPromotionError as exc:
        _raise_http(exc)


@router.post("/design-release-requests", status_code=201)
def create_design_release_request(
    body: ReleaseRequestCreate, request: Request, response: Response
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).create_release_request(body, actor=_actor(request))
    except DesignPromotionError as exc:
        _raise_http(exc)


@router.post("/design-release-requests/{release_request_id}/decision")
def decide_design_release(
    release_request_id: ReleaseRequestId,
    body: ReleaseDecisionRequest,
    request: Request,
    response: Response,
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).decide_release(
            release_request_id, body, actor=_actor(request)
        )
    except DesignPromotionError as exc:
        _raise_http(exc)


@router.post("/design-release-requests/{release_request_id}/publish")
def publish_design_release(
    release_request_id: ReleaseRequestId,
    body: PublishRequest,
    request: Request,
    response: Response,
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).publish_release(
            release_request_id, body, actor=_actor(request)
        )
    except DesignPromotionError as exc:
        _raise_http(exc)


@router.post("/design-release-requests/{release_request_id}/rollback")
def rollback_design_release(
    release_request_id: ReleaseRequestId,
    body: RollbackRequest,
    request: Request,
    response: Response,
) -> dict[str, Any]:
    response.headers["Cache-Control"] = "no-store"
    try:
        return _service(request).rollback_release(
            release_request_id, body, actor=_actor(request)
        )
    except DesignPromotionError as exc:
        _raise_http(exc)
